spectrochempy/analysis/mcrals.py

```diff
@@ -439,56 +496,53 @@ def _guess_profile(self, profile):
                 f"of X [{self._X_shape}]."
             )
 
-        # data array
-        Xdata = self._X.data
-
         # mask info
         if np.any(self._X_mask):
             masked_rows, masked_columns = self._get_masked_rc(self._X_mask)
 
         # make the profile
-        try:  # first try on concentration
-            # The data are validated in _C_validate()
-            # if it fails here due to shape mismatch, it goes to the except
-
-            Cdata = profile.copy()
-            n_components = Cdata.shape[1]
-            info_(f"Concentration profile initialized with {n_components} components")
-            # compute initial spectra (using the Xdata eventually masked
-            Stdata = np.linalg.lstsq(Cdata, Xdata, rcond=None)[0]
-            info_("Spectra profile computed")
+        if profile.shape[0] == self._X_shape[0]:
+            # this should be a concentration profile.
+            C = profile.copy()
+            self._n_components = C.shape[1]
+            info_(
+                f"Concentration profile initialized with {self._n_components} components"
+            )
+
+            # compute initial spectra (using X eventually masked)
+            # ... NonnegConc can be used by the solver, so must be validated:
+            proposal_nonnegConc = Meta()
+            proposal_nonnegConc.value = self.nonnegConc
+            self.nonnegConc = self._validate_nonnegConc(proposal_nonnegConc)
```

Comment:
If you simply write : 
   
     self.nonnegConc = self.nonnegConc

validation should occurs. Is-it not the case???

Reply:
Actually it is strange that it works with your trick, as the following 
lines:

    # ... Now solve
    St = self._solve_St(C)

use  `self.nonnegSpec`  not `self.nonnegConc`. 

For me, after replacing

    # ... NonnegConc can be used by the solver, so must be validated:
    proposal_nonnegConc = Meta()
    proposal_nonnegConc.value = self.nonnegConc
    self._validate_nonnegConc(proposal_nonnegConc)

by 
            
     self.nonnegSpec = self.nonnegSpec        
     # note Spec not Conc

and changing all `C_solver` by  `solverConc` 
it works. 

BUT actually the validation should already be done when reaching this part of the program. The problem is that it is done only at the end of the `_preprocess_as_Y_changed` function when line 772 is executed:

     self._Y_preprocessed = (C, St)

This fire the validation process  through the observe function  `_Y_preprocessed_change` but this is too late.

I think a good way to solve this is to replace the observe function by watching `_n_components` instead of `_Y_preprocessed` !

So replace (line 721)

    @tr.observe("_Y_preprocessed")
    def _Y_preprocessed_change(self, change):
        if self._n_components > 0: 

by

    @tr.observe("_n_components")
    def _n_components_change(self, change):
        if self._n_components > 0:

so that all validations are done early enough (as soon as `_n_component` is set).

Now you don't need anymore to do the validation inside the `_guess_profile` function.

Reply:
Perfect !  Thank you very much.